lib/spans/base-span-streamer.js

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2020 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const logger = require('../logger').child({ component: 'base-span-streamer' })
+const NAMES = require('../metrics/names').INFINITE_TRACING
+const SPAN_DROP_MSG_INTERVAL_MS = 30000
+const SPAN_DROP_MSG =
+  'Queue full, dropping spans. ' +
+  `Will not warn again for ${SPAN_DROP_MSG_INTERVAL_MS / 1000} seconds.`
+
+class BaseSpanStreamer {
+  constructor(licenseKey, connection, metrics, queueSize) {
+    this.stream = null
+    this.license_key = licenseKey
+    this.connection = connection
+    this.queue_size = queueSize
+    this.spans = []
+    this._metrics = metrics
+    this._writable = false
+
+    // 'connected' indicates a safely writeable stream.
+    // May still be mid-connect to gRPC server.
+    this.connection.on('connected', (stream) => {
+      logger.info('Span streamer connected')
+      this.stream = stream
+      this._writable = true
+      this.sendQueue()
+    })
+
+    this.connection.on('disconnected', () => {
+      logger.info('Span streamer disconnected')
+      this.stream = null
+      this._writable = false
+    })
+  }
+
+  addToQueue() {
+    throw new Error('addToQueue is not implemented')
+  }
```

Comment:
We can probably get rid of this dead code, unless you think we'll have other types of span streamers in the future and think we might forget to implement this method for them.

Reply:
If it isn't dead, then do you want a test that exercises it?

Reply:
sure I can add some unit tests for these methods that throw

Reply:
if we had typescript these would be in the interface definition 😄